File: code/programs/ruby/unix-tools/true_tool.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# true_tool.rb -- Do nothing, successfully
# =========================================
#
# === What This Program Does ===
#
# This is a reimplementation of the POSIX `true` utility. It does nothing
# and exits with a status code of 0 (success). That's it. No output, no
# side effects -- just a successful exit.
#
# === Why Does This Exist? ===
#
# `true` is one of the simplest Unix programs. It exists because shell
# scripts often need a command that always succeeds. Common uses:
#
#   - Infinite loops:    `while true; do ...; done`
#   - Default commands:  `PAGER="${PAGER:-true}"` (no-op if unset)
#   - Conditional chains: `true && echo "this always runs"`
#
# The POSIX standard requires that `true` accept (and ignore) any
# operands or options, but GNU coreutils adds `--help` and `--version`.
# Our implementation follows the GNU convention via CLI Builder.
#
# === How CLI Builder Powers This ===
#
# Even though `true` has no flags or arguments of its own, CLI Builder
# still provides `--help` and `--version` for free via the JSON spec.
# The spec declares no flags and no arguments -- CLI Builder handles
# the rest.

require "coding_adventures_cli_builder"

# ---------------------------------------------------------------------------
# Locate the JSON spec file.
# ---------------------------------------------------------------------------
# The spec file lives alongside this script. We resolve the path relative
# to this file's location so that the program works regardless of the
# user's current directory.

TRUE_SPEC_FILE = File.join(File.dirname(__FILE__), "true.json")

# ---------------------------------------------------------------------------
# Main entry point
# ---------------------------------------------------------------------------
# Parse arguments via CLI Builder, then exit successfully.
#
# The only interesting cases are --help and --version, which CLI Builder
# handles. For a normal invocation (no flags), we simply exit 0.

def true_main
  # --- Step 1: Parse arguments ---------------------------------------------
  # Hand the spec file and ARGV to CLI Builder. We prepend "true" to ARGV
  # because CLI Builder expects argv[0] to be the program name.
  begin
    result = CodingAdventures::CliBuilder::Parser.new(TRUE_SPEC_FILE, ["true"] + ARGV).parse
  rescue CodingAdventures::CliBuilder::ParseErrors => e
    # GNU `true` ignores all errors and always exits 0. However, we still
    # handle --help and --version via CLI Builder, so we parse first. If
    # parsing fails (e.g., truly malformed input), we exit 0 anyway --
    # that's what `true` does.
    e.errors.each { |err| warn "true: #{err.message}" }
    exit 0
  end

  # --- Step 2: Dispatch on result type -------------------------------------
  case result
  when CodingAdventures::CliBuilder::HelpResult
    puts result.text
    exit 0
  when CodingAdventures::CliBuilder::VersionResult
    puts result.version
    exit 0
  when CodingAdventures::CliBuilder::ParseResult
    # --- Step 3: Business logic --------------------------------------------
    # The business logic of `true` is beautifully simple: do nothing.
    # The program exits with status 0 (success) by falling through.
    exit 0
  end
end

# Only run main when this file is executed directly (not when required
# from tests). This is the Ruby equivalent of Python's
# `if __name__ == "__main__"`.
true_main if __FILE__ == $PROGRAM_NAME

File: code/programs/python/unix-tools/true_tool.py
"""true — do nothing, successfully.

=== What This Program Does ===

This is a reimplementation of the POSIX ``true`` utility. It does absolutely
nothing and exits with status code 0 (success). That's it. That's the whole
program.

=== Why Does This Exist? ===

At first glance, a program that does nothing seems pointless. But ``true`` is
a fundamental building block in shell scripting. It's used in:

1. **Infinite loops**: ``while true; do ... done``
2. **Default success commands**: ``COMMAND || true`` (ignore failures)
3. **Placeholder commands**: When a script needs a command that always succeeds

For example, in a Makefile you might write::

    clean:
        rm -f *.o || true

The ``|| true`` ensures that the ``clean`` target succeeds even if there are
no ``.o`` files to remove. Without it, ``rm`` would fail and ``make`` would
report an error.

=== CLI Builder Integration ===

Even though ``true`` takes no meaningful arguments, we still wire it through
CLI Builder so that ``true --help`` and ``true --version`` work correctly.
Any other arguments are silently ignored (per POSIX behavior), but the JSON
spec doesn't define any flags or arguments beyond the builtins.
"""

from __future__ import annotations

import sys
from pathlib import Path

# ---------------------------------------------------------------------------
# Locate the JSON spec file.
# ---------------------------------------------------------------------------
# The spec file lives alongside this script. We resolve the path relative
# to this file's location so that the program works regardless of the
# user's current directory.

SPEC_FILE = str(Path(__file__).parent / "true.json")


def main() -> None:
    """Entry point: parse args via CLI Builder, then exit 0.

    The ``true`` utility must always exit with status 0, regardless of
    what arguments are passed. The only exceptions are ``--help`` and
    ``--version``, which print their respective output and then exit 0
    as well.
    """
    # Import CLI Builder. This import will fail if the package is not
    # installed — see the BUILD file for how dependencies are set up.
    from cli_builder import HelpResult, ParseErrors, Parser, ParseResult, VersionResult

    # --- Step 1: Parse arguments -------------------------------------------
    # Hand the spec file and sys.argv to CLI Builder. Even though true
    # ignores all arguments, we still parse them so --help and --version
    # work correctly.
    try:
        result = Parser(SPEC_FILE, sys.argv).parse()
    except ParseErrors:
        # POSIX true ignores all errors — it always succeeds.
        # But we still exit 0 because that's what true does.
        raise SystemExit(0) from None

    # --- Step 2: Dispatch on result type -----------------------------------
    # CLI Builder returns one of:
    #   - HelpResult:    user passed --help
    #   - VersionResult: user passed --version
    #   - ParseResult:   normal invocation; nothing to do

    if isinstance(result, HelpResult):
        print(result.text)
        raise SystemExit(0)

    if isinstance(result, VersionResult):
        print(result.version)
        raise SystemExit(0)

    # --- Step 3: Business logic --------------------------------------------
    # The business logic of ``true`` is: do nothing. Exit successfully.
    # This is the simplest possible program — a no-op that returns 0.

    assert isinstance(result, ParseResult)
    raise SystemExit(0)


if __name__ == "__main__":
    main()
